### PR TITLE
MWP PredictMask()

### PR DESCRIPTION
**Pull request by EricFillion**
**Changes:** 

- **Completed predict_mask() method.** 
- Changed default model for MWP to "distilbert-base-uncased"
- Added test cases 
- Added Not Implemented errors 

**Testing:** 
- found under tests.test_mwp()

**Notes**
- I think it's best if we hold off on adding the MWP training functionality for V2.0.0. Perhaps we'll add it for V2.1.0. 
- The FillMaskPipeline allows for an input of a list of strings, where each string contains a mask to be predicted. I blocked this functionality for the sake of simplicity and to keep the documentation more concise. We can perhaps unblock it for a later release. 
- We can add functionality for predicting multiple masks within a string in a later release.

### CODE REVIEW

File: happytransformer/happy_word_prediction.py [hunk 1]
```diff
@@ -15,36 +13,56 @@
...
+    def predict_mask(self, text, targets=None, top_k=1):
+        """
+        :param text: A string that contains the model's mask token
+        :param targets: Optional. A list of strings of potential answers.
+        All other answers will be ignored
+        :param top_k: number of results. Default is 1
+        :return: If top_k ==1: a dictionary with the keys "score" and "token_str"
+                if  top_k >1: a list of dictionaries described above in order by score
+        """
+        if isinstance(text, list):
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:46]: could we replace this with `if not isinstance(text, str)`? would cover more cases

File: happytransformer/happy_word_prediction.py [hunk 2]
```diff
@@ -15,36 +13,56 @@
...
-        self._trainer.train(input_filepath=input_filepath, args=args)
+        raise NotImplementedError("train() is currently not available")
 
     def eval(self, input_filepath):
-        return self._trainer.eval(input_filepath=input_filepath)
+        raise NotImplementedError("eval() is currently not available")
 
     def test(self, input_filepath):
-        return self._trainer.test(input_filepath=input_filepath,)
+        # self.logger.error("test() is currently not available")
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:67]: remove comment

File: tests/test_mwp.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+from happytransformer import HappyWordPrediction
+
+
+def test_mwp_basic():
+    happy_mwp = HappyWordPrediction()
+    result = happy_mwp.predict_mask(
+        "Please pass the salt and [MASK]",
+    )
+    answer = {'score': 0.2664579749107361, 'token_str': 'pepper'}
```
Comment (ted537) [inline tests/test_mwp.py:9]: if we're rewriting for v2 can we use named tuples?

File: happytransformer/happy_word_prediction.py [hunk 4]
```diff
@@ -15,36 +13,56 @@
...
+        self._pipeline = FillMaskPipeline(model=model, tokenizer=tokenizer, device=device_number)
+        self._trainer = MWPTrainer(model, model_type, tokenizer, self._device, self.logger)
+
+    def predict_mask(self, text, targets=None, top_k=1):
+        """
+        :param text: A string that contains the model's mask token
+        :param targets: Optional. A list of strings of potential answers.
+        All other answers will be ignored
+        :param top_k: number of results. Default is 1
+        :return: If top_k ==1: a dictionary with the keys "score" and "token_str"
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:43]: returning different types seems odd, especially when based on a non-boolean parameter

File: happytransformer/happy_word_prediction.py [hunk 5]
```diff
@@ -15,36 +13,56 @@
...
+        result = self._pipeline(text, targets=targets, top_k=top_k)
+        if top_k == 1:
+            result = result[0]
+            del result['sequence']
+            del result['token']
+        else:
+            for answer in result:
+                del answer['sequence']
+                del answer['token']
+        return result
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:58]: returning a dict with what we want explicitly would be cleaner than deleting things

**Review by ted537 — CHANGES_REQUESTED** [review]

File: tests/test_mwp.py (hunk 3 above)
Comment (EricFillion) [reply tests/test_mwp.py:9]: yes good call

File: happytransformer/happy_word_prediction.py (hunk 1 above)
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:46]: Will do

File: happytransformer/happy_word_prediction.py (hunk 2 above)
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:67]: Will do

File: happytransformer/happy_word_prediction.py (hunk 4 above)
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:43]: I agree. Do you think we should just always return a list?

Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:43]: A list of named tuples

Comment (ted537) [reply happytransformer/happy_word_prediction.py:43]: That would make the most sense I think :)

**Comment by EricFillion** [comment]
I added your suggestions. I also renamed a few things to remove the M in MWP. More concise this way.

File: happytransformer/happy_word_prediction.py [hunk 6]
```diff
@@ -7,9 +7,11 @@
 
 )
 import torch
-
+import collections
 from happytransformer.happy_transformer import HappyTransformer
-from happytransformer.mwp.trainer import  MWPTrainer
+from happytransformer.mwp.trainer import WPTrainer
+
+WPOutput = collections.namedtuple("WPOutput", ["token_str", "score"])
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:14]: I think `WordPredictionResult` would be a more descriptive name. "Output" sometimes has broader connotations, whereas Result is more likely to be interpreted by users as a single item in my opinion.

**Review by ted537 — CHANGES_REQUESTED** [review]

File: happytransformer/happy_word_prediction.py (hunk 4 above)
Comment (ted537) [reply happytransformer/happy_word_prediction.py:45]: docstring is still out of date

Comment (EricFillion) [reply happytransformer/happy_word_prediction.py]: Updated

File: happytransformer/happy_word_prediction.py [hunk 7]
```diff
@@ -43,19 +45,16 @@ def predict_mask(self, text, targets=None, top_k=1):
...
-            for answer in result:
-                del answer['sequence']
-                del answer['token']
-        return result
+        results = list()
+        for answer in result:
+            result = WPOutput(token_str=answer["token_str"], score=answer["score"])
+            results.append(result)
+
+        return results
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py]: Updated

File: happytransformer/happy_word_prediction.py (hunk 6 above)
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py]: Updated

**Review by ted537 — APPROVED** [review]
